src/rate_limit.rs

```diff
@@ -0,0 +1,239 @@
+            self.update();
+        };
+
+        self.hit(peer_key)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::tests::test_utils::pubkey;
+    use core::ops::SubAssign;
+    use mockall::mock;
+    use tokio::time::{Duration, Instant};
+
+    const TEST_COUNT: u8 = 2;
+    const TEST_FREQUENCY: Duration = Duration::from_secs(1);
+
+    mock! {
+        FixedClock{}
+
+        impl Clock for FixedClock{
+            fn now(&self) -> Instant;
+        }
+    }
+
+    #[test]
+    fn test_peer_connection() {
+        let pk_0 = pubkey(0);
+        let pk_1 = pubkey(1);
+
+        let mut clock = MockFixedClock::new();
+        let start_time = Instant::now();
+        clock.expect_now().returning(move || start_time);
+
+        // Assert that we're set up with our original peer.
+        let mut rate_limiter =
+            TokenLimiter::new(vec![pk_0].into_iter(), TEST_COUNT, TEST_FREQUENCY, clock);
+        assert_eq!(rate_limiter.peers(), vec![pk_0]);
+
+        // Connect a new peer and assert that both are reported.
+        rate_limiter.peer_connected(pk_1);
+        assert_eq!(rate_limiter.peers().sort(), vec![pk_0, pk_1].sort());
+
+        // Disconnect our original peer and assert that it's no longer listed.
+        rate_limiter.peer_disconnected(pk_0);
+        assert_eq!(rate_limiter.peers(), vec![pk_1]);
+    }
+
+    #[test]
+    fn test_rate_limiting() {
+        let pk_0 = pubkey(0);
+        let pk_1 = pubkey(1);
+
+        let mut clock = MockFixedClock::new();
+
+        // TODO: use constant value for mocking.
+        let start_time = Instant::now();
+        clock.expect_now().returning(move || start_time);
+
+        // Assert that we're set up with our original peer.
+        let mut rate_limiter =
+            TokenLimiter::new(vec![pk_0].into_iter(), TEST_COUNT, TEST_FREQUENCY, clock);
+        assert_eq!(rate_limiter.peers(), vec![pk_0]);
+
+        // Exhaust our allowed call count in this period, then assert that we're no longer allowed to query further.
+        for _ in 0..TEST_COUNT {
+            assert!(rate_limiter.query_peer(pk_0));
+        }
+        assert!(!rate_limiter.query_peer(pk_0));
+
+        // Connect a new peer, use some calls, then flip our connection off and on and assert that they aren't allowed
+        // more calls by virtue of having disconnected.
+        rate_limiter.peer_connected(pk_1);
+        assert!(rate_limiter.query_peer(pk_1));
+
+        rate_limiter.peer_disconnected(pk_1);
+        rate_limiter.peer_connected(pk_1);
+
+        assert!(rate_limiter.query_peer(pk_1));
+        assert!(!rate_limiter.query_peer(pk_1));
+
+        // Disconnect one peer so that they'll be cleaned up and update the rate limiter's time so that we'll refresh
+        // our buckets.
+        rate_limiter.peer_disconnected(pk_1);
+
+        // Update our clock to a time which reflects that we need an update.
+        rate_limiter.last_update.sub_assign(TEST_FREQUENCY);
```

Comment:
Not very happy with these tests - the mocked clock isn't easily clone-able, so we can't update it once it's passed to `TokenLimiter` (and even if we passed `&clock` to `TokenLimiter`, we still can't have mutable borrows). 

If these pass basic human diginity, would like to ask that we put them in as is and then clean up later on!

Reply:
Sounds good to me :+1: